libs/langchain_v1/langchain/agents/middleware/human_in_the_loop.py

```diff
@@ -1,135 +1,201 @@
 """Human in the loop middleware."""
 
-from typing import Any
-
-from langgraph.prebuilt.interrupt import (
-    ActionRequest,
-    HumanInterrupt,
-    HumanInterruptConfig,
-    HumanResponse,
-)
+from typing import Any, Literal
+
+from langchain_core.messages import AIMessage, ToolCall, ToolMessage
 from langgraph.types import interrupt
+from typing_extensions import NotRequired, TypedDict
 
-from langchain.agents.middleware._utils import _generate_correction_tool_messages
 from langchain.agents.middleware.types import AgentMiddleware, AgentState
 
-ToolInterruptConfig = dict[str, HumanInterruptConfig]
+
+class ActionRequest(TypedDict):
+    """Request for human feedback.
+
+    Attributes:
+        message: Human-readable message describing what action is needed
+        args: Args associated with the action (tool kwargs, for example)
+        allowed_actions: List of allowed action types, can be one of "approve", "reject", "edit"
+        description: Optional detailed description
+
+        TODO: add args_schema to specify the allowed schema for the response args
+    """
+
+    message: str
+    args: dict
+    allowed_actions: list[Literal["approve", "reject", "edit"]]
+    description: str | None
+
+
+class ApproveResponse(TypedDict):
+    """Response when a human approves the action."""
+
+    action: Literal["approve"]
```

Comment:
name collision in `action`. `action` refers the action taken by the reviewer / HIL as well as to the action taken by the agent.